tests/qos/qos_sai_base.py

```diff
@@ -1070,3 +1053,295 @@ def resetWatermark(
+        """
+        Returns list of port index and IP address for a given ASIC
+        """
+
+        dut_port_ips = dict()
+
+        for iface, addr in ifaces.items():
+            if iface.startswith("Ethernet"):
+                portIndex = mg_facts["minigraph_ptf_indices"][iface]
+            elif iface.startswith("PortChannel"):
+                portName = mg_facts["minigraph_portchannels"][iface]["members"][0]
+                portIndex = mg_facts["minigraph_ptf_indices"][portName]
+
+            dut_port_ips.update({
+                portIndex: {
+                    "ipv4": addr["peer_ipv4"],
+                    "bgp_neighbor": addr["bgp_neighbor"]
+                }
+            })
+
+        return {asic_index: dut_port_ips}
+
+    def get_backend_ip_ifs(self, duthost, frontend_asic):
+        """
+        On a frontend ASIC return a dict of interfaces with
+        backend ASIC names
+        """
+        pytest_assert(
+            frontend_asic in duthost.get_frontend_asic_ids(),
+            "{} is not frontend ASIC ID".format(frontend_asic)
+        )
+
+        ip_ifs = duthost.asic_instance(
+            frontend_asic
+        ).show_ip_interface()["ansible_facts"]["ip_interfaces"]
+
+        # Find backend interface names
+        return {intf: ip["bgp_neighbor"].lower() for intf, ip in ip_ifs.items()
+                if ip["bgp_neighbor"].lower().startswith("asic")}
+
+    def check_v4route_backend_nhop(self, duthost, frontend_asic, route):
+        """
+        On frontend ASIC Check if v4 address has at least one backend
+        ASIC nexthop
+
+        Returns:
+          False if not nexthops with backend ASICs
+        """
+        cmd = 'vtysh -n {} -c "show ip route {} json"'.format(
+            frontend_asic, route
+        )
+        result = duthost.command(cmd)
+        pytest_assert(result["rc"] == 0, cmd)
+        route_info = json.loads(result["stdout"])
+        nhop = route_info[route_info.keys().pop()][0]
+
+        nhop_ifs = {x["interfaceName"] for x in nhop["nexthops"]}
+        backend_ifs = set(self.get_backend_ip_ifs(
+            duthost, frontend_asic).keys()
+        )
+
+        return len(nhop_ifs.intersection(backend_ifs))
+
+    def backend_ip_if_admin_state(
+        self, duthost, test_asic, frontend_asic, admin_state
+    ):
+        """
+        On a frontend ASIC bring down ports (channels) towards backend ASICs
+        other than the ASIC under test, so that traffic always goes via
+        backend ASIC under test
+        """
+
+        def is_intf_status(asic, intf, oper_state):
+            intf_status = duthost.asic_instance(asic).show_interface(
+                command="status", include_internal_intfs=True
+            )["ansible_facts"]["int_status"]
+            if intf_status[intf]["oper_state"] == oper_state:
+                return True
+            return False
+
+        asic_ns = "asic{}".format(test_asic)
+        cmd_str = "sudo config interface -n asic{} {} {}"
```

Comment:
use this function in sonicAsic https://github.com/Azure/sonic-mgmt/blob/a5ea4bf6b77438dfe1e6d3f413c055e7c04470c2/tests/common/devices/sonic_asic.py#L448  iso using the cli directly ?